### DISCUSSION TITLE
IIS installation

### DISCUSSION THREAD

**Discussion by Eaclean** [Leantime/leantime#1342] [Q&A] [answered]
Hi,

I have setup leantime on IIS using PHP8 but I'm having some issues.

When I visit the page /install url I am getting a list of errors and nothing seams to render.

Does anyone have any idea what's going on?

Here are the list of errors.

Thanks

Deprecated: Creation of dynamic property leantime\core\environment::$dotenv is deprecated in E:\leantime\app\core\class.environment.php on line 23

Deprecated: Creation of dynamic property leantime\core\environment::$yaml is deprecated in E:\leantime\app\core\class.environment.php on line 26

Deprecated: Creation of dynamic property leantime\core\environment::$sitename is deprecated in E:\leantime\app\core\class.environment.php on line 32

Deprecated: Creation of dynamic property leantime\core\environment::$language is deprecated in E:\leantime\app\core\class.environment.php on line 33

Deprecated: Creation of dynamic property leantime\core\environment::$logoPath is deprecated in E:\leantime\app\core\class.environment.php on line 34

Deprecated: Creation of dynamic property leantime\core\environment::$printLogoURL is deprecated in E:\leantime\app\core\class.environment.php on line 35

Deprecated: Creation of dynamic property leantime\core\environment::$appUrl is deprecated in E:\leantime\app\core\class.environment.php on line 36

Deprecated: Creation of dynamic property leantime\core\environment::$defaultTheme is deprecated in E:\leantime\app\core\class.environment.php on line 37

Deprecated: Creation of dynamic property leantime\core\environment::$primarycolor is deprecated in E:\leantime\app\core\class.environment.php on line 38

Deprecated: Creation of dynamic property leantime\core\environment::$secondarycolor is deprecated in E:\leantime\app\core\class.environment.php on line 39

Deprecated: Creation of dynamic property leantime\core\environment::$debug is deprecated in E:\leantime\app\core\class.environment.php on line 40

Deprecated: Creation of dynamic property leantime\core\environment::$defaultTimezone is deprecated in E:\leantime\app\core\class.environment.php on line 41

Deprecated: Creation of dynamic property leantime\core\environment::$enableMenuType is deprecated in E:\leantime\app\core\class.environment.php on line 42

Deprecated: Creation of dynamic property leantime\core\environment::$keepTheme is deprecated in E:\leantime\app\core\class.environment.php on line 43

Deprecated: Creation of dynamic property leantime\core\environment::$appUrlRoot is deprecated in E:\leantime\app\core\class.environment.php on line 46

Deprecated: Creation of dynamic property leantime\core\environment::$dbHost is deprecated in E:\leantime\app\core\class.environment.php on line 49

Deprecated: Creation of dynamic property leantime\core\environment::$dbUser is deprecated in E:\leantime\app\core\class.environment.php on line 50

Deprecated: Creation of dynamic property leantime\core\environment::$dbPassword is deprecated in E:\leantime\app\core\class.environment.php on line 51

Deprecated: Creation of dynamic property leantime\core\environment::$dbDatabase is deprecated in E:\leantime\app\core\class.environment.php on line 52

Deprecated: Creation of dynamic property leantime\core\environment::$dbPort is deprecated in E:\leantime\app\core\class.environment.php on line 53

Deprecated: Creation of dynamic property leantime\core\environment::$userFilePath is deprecated in E:\leantime\app\core\class.environment.php on line 56

Deprecated: Creation of dynamic property leantime\core\environment::$useS3 is deprecated in E:\leantime\app\core\class.environment.php on line 57

Deprecated: Creation of dynamic property leantime\core\environment::$sessionpassword is deprecated in E:\leantime\app\core\class.environment.php on line 69

Deprecated: Creation of dynamic property leantime\core\environment::$sessionExpiration is deprecated in E:\leantime\app\core\class.environment.php on line 70

Deprecated: Creation of dynamic property leantime\core\environment::$email is deprecated in E:\leantime\app\core\class.environment.php on line 73

Deprecated: Creation of dynamic property leantime\core\environment::$useSMTP is deprecated in E:\leantime\app\core\class.environment.php on line 74

Deprecated: Creation of dynamic property leantime\core\environment::$useLdap is deprecated in E:\leantime\app\core\class.environment.php on line 87

Warning: ini_set(): Session ini settings cannot be changed after headers have already been sent in E:\leantime\config\appSettings.php on line 55

Warning: ini_set(): Session ini settings cannot be changed after headers have already been sent in E:\leantime\config\appSettings.php on line 56

Warning: ini_set(): Session ini settings cannot be changed after headers have already been sent in E:\leantime\config\appSettings.php on line 57

Warning: ini_set(): Session ini settings cannot be changed after headers have already been sent in E:\leantime\config\appSettings.php on line 58

Deprecated: Using ${var} in strings is deprecated, use {$var} instead in E:\leantime\app\core\class.application.php on line 175

Deprecated: Using ${var} in strings is deprecated, use {$var} instead in E:\leantime\app\core\class.application.php on line 175

Warning: ini_set(): Session ini settings cannot be changed after headers have already been sent in E:\leantime\app\core\class.session.php on line 44

Warning: ini_set(): Session ini settings cannot be changed after headers have already been sent in E:\leantime\app\core\class.session.php on line 45

Warning: session_name(): Session name cannot be changed after headers have already been sent in E:\leantime\app\core\class.session.php on line 72

Warning: session_id(): Session ID cannot be changed after headers have already been sent in E:\leantime\app\core\class.session.php on line 73

Warning: session_start(): Session cannot be started after headers have already been sent in E:\leantime\app\core\class.session.php on line 74

Warning: Cannot modify header information - headers already sent by (output started at E:\leantime\app\core\class.environment.php:69) in E:\leantime\app\core\class.session.php on line 76

Deprecated: Creation of dynamic property leantime\core\language::$config is deprecated in E:\leantime\app\core\class.language.php on line 71

Deprecated: Creation of dynamic property leantime\core\language::$themeCore is deprecated in E:\leantime\app\core\class.language.php on line 73

Deprecated: Creation of dynamic property leantime\core\language::$theme is deprecated in E:\leantime\app\core\class.language.php on line 74

Deprecated: Creation of dynamic property leantime\domain\services\auth::$language is deprecated in E:\leantime\app\domain\auth\services\class.auth.php on line 138

Deprecated: Creation of dynamic property leantime\domain\services\auth::$settingsRepo is deprecated in E:\leantime\app\domain\auth\services\class.auth.php on line 139

Deprecated: Creation of dynamic property leantime\core\template::$theme is deprecated in E:\leantime\app\core\class.template.php on line 86

Deprecated: Creation of dynamic property leantime\domain\repositories\users::$config is deprecated in E:\leantime\app\domain\users\repositories\class.users.php on line 67

Deprecated: Creation of dynamic property leantime\domain\repositories\auth::$userService is deprecated in E:\leantime\app\domain\auth\repositories\class.auth.php on line 139

Deprecated: Creation of dynamic property leantime\domain\repositories\users::$config is deprecated in E:\leantime\app\domain\users\repositories\class.users.php on line 67

Deprecated: Creation of dynamic property leantime\domain\repositories\auth::$userRepo is deprecated in E:\leantime\app\domain\auth\repositories\class.auth.php on line 140

Deprecated: Creation of dynamic property leantime\domain\services\auth::$authRepo is deprecated in E:\leantime\app\domain\auth\services\class.auth.php on line 140

Deprecated: Creation of dynamic property leantime\domain\repositories\users::$config is deprecated in E:\leantime\app\domain\users\repositories\class.users.php on line 67

Deprecated: Creation of dynamic property leantime\domain\services\auth::$userRepo is deprecated in E:\leantime\app\domain\auth\services\class.auth.php on line 141

Deprecated: Creation of dynamic property leantime\core\template::$theme is deprecated in E:\leantime\app\core\class.template.php on line 86

Deprecated: Creation of dynamic property leantime\domain\repositories\files::$db is deprecated in E:\leantime\app\domain\files\repositories\class.files.php on line 17

Deprecated: Creation of dynamic property leantime\domain\services\projects::$filesRepository is deprecated in E:\leantime\app\domain\projects\services\class.projects.php on line 37

Deprecated: Creation of dynamic property leantime\domain\services\notifications::$db is deprecated in E:\leantime\app\domain\notifications\services\class.notifications.php on line 21

Deprecated: Creation of dynamic property leantime\domain\repositories\notifications::$db is deprecated in E:\leantime\app\domain\notifications\repositories\class.notifications.php on line 22

Deprecated: Creation of dynamic property leantime\domain\services\notifications::$notificationsRepo is deprecated in E:\leantime\app\domain\notifications\services\class.notifications.php on line 22

Deprecated: Creation of dynamic property leantime\domain\repositories\users::$config is deprecated in E:\leantime\app\domain\users\repositories\class.users.php on line 67

Deprecated: Creation of dynamic property leantime\domain\services\notifications::$userRepository is deprecated in E:\leantime\app\domain\notifications\services\class.notifications.php on line 23

Deprecated: Creation of dynamic property leantime\domain\services\notifications::$language is deprecated in E:\leantime\app\domain\notifications\services\class.notifications.php on line 24

Deprecated: Creation of dynamic property leantime\core\template::$theme is deprecated in E:\leantime\app\core\class.template.php on line 86

Deprecated: Creation of dynamic property leantime\domain\services\reports::$tpl is deprecated in E:\leantime\app\domain\reports\services\class.reports.php on line 20

Deprecated: Creation of dynamic property leantime\domain\repositories\sprints::$db is deprecated in E:\leantime\app\domain\sprints\repositories\class.sprints.php on line 20

Deprecated: Creation of dynamic property leantime\domain\repositories\reports::$db is deprecated in E:\leantime\app\domain\reports\repositories\class.reports.php on line 22

Deprecated: Creation of dynamic property leantime\domain\services\reports::$settings is deprecated in E:\leantime\app\domain\reports\services\class.reports.php on line 24

Deprecated: Creation of dynamic property leantime\domain\services\reports::$appSettings is deprecated in E:\leantime\app\domain\reports\services\class.reports.php on line 25

Warning: Cannot modify header information - headers already sent by (output started at E:\leantime\app\core\class.environment.php:69) in E:\leantime\app\core\class.application.php on line 175

Warning: Cannot modify header information - headers already sent by (output started at E:\leantime\app\core\class.environment.php:69) in E:\leantime\app\core\class.application.php on line 175

Warning: Cannot modify header information - headers already sent by (output started at E:\leantime\app\core\class.environment.php:69) in E:\leantime\app\core\class.application.php on line 175

Warning: Cannot modify header information - headers already sent by (output started at E:\leantime\app\core\class.environment.php:69) in E:\leantime\app\core\class.frontcontroller.php on line 301

**Comment by marcelfolaron** [comment]
Hi,

Seems like you are on PHP 8.2 ? I haven't seen those deprecation notices
before.
We haven't done a lot of testing in 8.2 yet. The errors should go away if
you turn off deprecation warnings.
error_reporting = E_ALL & ~E_DEPRECATED & ~E_STRICT

**Reply by Eaclean** [reply]
Thanks very much for your response.

This has removed all of those errors.

I am now left with a shorter error message which by the looks of it contains a "Fatal Error" regarding pdo mysql. I have the php_pdo_mysql.dll enabled and I can see it in the phpmyinfo file so not sure what I could do.

Any ideas?

PHP Deprecated: Using ${var} in strings is deprecated, use {$var} instead in E:\leantime\app\core\class.application.php on line 175 PHP Deprecated: Using ${var} in strings is deprecated, use {$var} instead in E:\leantime\app\core\class.application.php on line 175 PHP Warning: session_start(): open(C:\Windows\TEMP\sess_418ac069119b5bff579a4808cd45721e7f476c48-8786902e01a6f78a917b28a0cb2e0dd6415cc175, O_RDWR) failed: Permission denied (13) in E:\leantime\app\core\class.session.php on line 74 PHP Warning: session_start(): Failed to read session data: files (path: ) in E:\leantime\app\core\class.session.php on line 74 PHP Deprecated: Creation of dynamic property leantime\core\language::$config is deprecated in E:\leantime\app\core\class.language.php on line 71 PHP Fatal error: Uncaught Error: Undefined constant PDO::MYSQL_ATTR_INIT_COMMAND in E:\leantime\app\core\class.db.php:74 Stack trace: #0 E:\leantime\app\core\class.db.php(104): leantime\core\db->__construct() #1 E:\leantime\app\domain\setting\repositories\class.setting.php(26): leantime\core\db::getInstance() #2 E:\leantime\app\core\class.language.php(72): leantime\domain\repositories\setting->__construct() #3 E:\leantime\app\core\class.language.php(158): leantime\core\language->__construct() #4 E:\leantime\app\domain\auth\services\class.auth.php(138): leantime\core\language::getInstance() #5 E:\leantime\app\domain\auth\services\class.auth.php(151): leantime\domain\services\auth->__construct('418ac069119b5bf...') #6 E:\leantime\app\core\class.application.php(40): leantime\domain\services\auth::getInstance('418ac069119b5bf...') #7 E:\leantime\public\index.php(24): leantime\core\application->__construct() #8 {main} thrown in E:\leantime\app\core\class.db.php on line 74

**Reply by marcelfolaron** [reply]
I see 2 things going on here:
1. PHP can't write to the session directory: PHP Warning: session_start(): open(C:\Windows\TEMP\sess_418ac069119b5bff579a4808cd45721e7f476c48-8786902e01a6f78a917b28a0cb2e0dd6415cc175, O_RDWR) failed: Permission denied 

2. There is still something not quite right with PDO. Is the pdo extension in the ini file commented out?: extension=php_pdo_mysql.dll on Windows
PHP Fatal error: Uncaught Error: Undefined constant PDO::MYSQL_ATTR_INIT_COMMAND

**Reply by Eaclean** [reply]
Hi,

Thanks for your help.

I have now got the permissions on the folder sorted and the PDO extension activated and working.

Next hurdle I have is IIS url rewrite settings.

Do you have a standard set of rules that can be used in the web.config file?

I did try these rules but I just get an error too many redirects message. The only change I made was remove the "s" from "https" below as it's our internal network, without https.

It jumps between 

/install/update

and 

/auth/login?redirect=%2install%2update


```
<rewrite>
            <rules>
                <rule name="Imported Rule 1" stopProcessing="true">
                    <match url=".*" ignoreCase="false" />
                    <conditions logicalGrouping="MatchAll">
                        <add input="{HTTP_X_FORWARDED_PROTO}" pattern="^http$" ignoreCase="false" />
                    </conditions>
                    <action type="Redirect" url="http://{HTTP_HOST}{URL}" redirectType="Permanent" />
                </rule>
                <rule name="Imported Rule 2">
                    <match url="^resetPassword$" ignoreCase="false" />
                    <action type="Rewrite" url="/index.php?resetPassword=true" appendQueryString="false" />
                </rule>
                <rule name="Imported Rule 3">
                    <match url="^resetPassword/([^/\.]+)/?$" ignoreCase="false" />
                    <action type="Rewrite" url="/index.php?resetPassword=true&amp;hash={R:1}" appendQueryString="false" />
                </rule>
                <rule name="Imported Rule 4">
                    <match url="^install$" ignoreCase="false" />
                    <action type="Rewrite" url="/index.php?install=true" appendQueryString="false" />
                </rule>
                <rule name="Imported Rule 5">
                    <match url="^install/([^/\.]+)/?$" ignoreCase="false" />
                    <action type="Rewrite" url="/index.php?install=true" appendQueryString="false" />
                </rule>
                <rule name="Imported Rule 6">
                    <match url="^update" ignoreCase="false" />
                    <action type="Rewrite" url="/index.php?update=true" appendQueryString="false" />
                </rule>
                <rule name="Imported Rule 7">
                    <match url="^update/([^/\.]+)/?$" ignoreCase="false" />
                    <action type="Rewrite" url="/index.php?update=true" appendQueryString="false" />
                </rule>
                <rule name="Imported Rule 8">
                    <match url="^$" ignoreCase="false" />
                    <action type="Rewrite" url="/index.php?act=dashboard.show" appendQueryString="false" />
                </rule>
                <rule name="Imported Rule 9">
                    <match url="^([^/\.]+)/([^/\.]+)/?$" ignoreCase="false" />
                    <action type="Rewrite" url="/index.php?act={R:1}.{R:2}" appendQueryString="true" />
                </rule>
                <rule name="Imported Rule 10">
                    <match url="^([^/\.]+)/([^/\.]+)/([^/\.]+)/?$" ignoreCase="false" />
                    <action type="Rewrite" url="/index.php?act={R:1}.{R:2}&amp;id={R:3}" appendQueryString="true" />
                </rule>
            </rules>
        </rewrite>
```

**Comment by Eaclean** [comment]
Hi,

I have the appUrl set to http://*********** so it's not that that's the issue. 

Are the url rewrite lines above the latest/standard rules?

Thanks

**Comment by marcelfolaron** [comment]
Hi,

There is a mod_rewrite module you can install to utilize htaccess
https://www.iis.net/downloads/microsoft/url-rewrite

**Comment by marcelfolaron** [comment, marked as answer]
There is also a htaccess to web.config converter. I don't know how well it works so this is untested:

```
<rule name="rule 1B">
	<match url="^/?$"  />
	<action type="Rewrite" url="/index.php?act=dashboard.show"  />
</rule>
<rule name="rule 2B">
	<match url="^([^/\.]+)/?$"  />
	<action type="Rewrite" url="/index.php?act={R:1}"  appendQueryString="true" />
</rule>
<rule name="rule 3B">
	<match url="^([^/\.]+)/([^/\.]+)/?$"  />
	<action type="Rewrite" url="/index.php?act={R:1}.{R:2}"  appendQueryString="true" />
</rule>
<rule name="rule 4B">
	<match url="^([^/\.]+)/([^/\.]+)/([^/\.]+)/?$"  />
	<action type="Rewrite" url="/index.php?act={R:1}.{R:2}&amp;id={R:3}"  appendQueryString="true" />
</rule>
``` 

Let me know if it works so I can add it to the repo
https://cbsa.com.br/tools/online-convert-htaccess-to-web-config.aspx#result

**Comment by Eaclean** [comment]
Thanks.

The above rewrite rules have worked, I have successfully installed Leantime and can see the database tables etc.

Next issue is after logging in I'm getting an error when going to any of the pages as shown below, any ideas?

Thanks
![Capture](https://user-images.githubusercontent.com/123482922/214854710-48f72d56-155d-44e9-bc86-e7392e093a72.PNG)

**Comment by marcelfolaron** [comment]
Can you check if there are any errors in logs/error.log ?

**Reply by Eaclean** [reply]
Hi,

This is the logs from the most recent request.

[26-Jan-2023 08:01:28 America/Los_Angeles] Intervention\Image\Exception\MissingDependencyException: PHP Fileinfo extension must be installed/enabled to use Intervention Image. in E:\leantime\vendor\intervention\image\src\Intervention\Image\ImageManager.php:137
Stack trace:
#0 E:\leantime\vendor\intervention\image\src\Intervention\Image\ImageManager.php(27): Intervention\Image\ImageManager->checkRequirements()
#1 E:\leantime\vendor\lasserafn\php-initial-avatar-generator\src\InitialAvatar.php(77): Intervention\Image\ImageManager->__construct()
#2 E:\leantime\vendor\lasserafn\php-initial-avatar-generator\src\InitialAvatar.php(69): LasseRafn\InitialAvatarGenerator\InitialAvatar->setupImageManager()
#3 E:\leantime\app\domain\users\repositories\class.users.php(567): LasseRafn\InitialAvatarGenerator\InitialAvatar->__construct()
#4 E:\leantime\app\domain\users\services\class.users.php(29): leantime\domain\repositories\users->getProfilePicture()
#5 E:\leantime\app\domain\auth\controllers\class.loginInfo.php(33): leantime\domain\services\users->getProfilePicture()
#6 E:\leantime\app\core\class.controller.php(63): leantime\domain\controllers\loginInfo->run()
#7 E:\leantime\app\core\class.controller.php(33): leantime\core\controller->executeActions()
#8 E:\leantime\app\core\class.frontcontroller.php(173): leantime\core\controller->__construct()
#9 E:\leantime\app\core\class.frontcontroller.php(233): leantime\core\frontcontroller::executeAction()
#10 E:\leantime\app\domain\menu\templates\headMenu.tpl.php(212): leantime\core\frontcontroller::includeAction()
#11 E:\leantime\app\core\class.template.php(259): require_once('...')
#12 E:\leantime\app\core\class.template.php(319): leantime\core\template->display()
#13 E:\leantime\app\domain\menu\controllers\class.headMenu.php(36): leantime\core\template->displayPartial()
#14 E:\leantime\app\core\class.controller.php(63): leantime\domain\controllers\headMenu->run()
#15 E:\leantime\app\core\class.controller.php(33): leantime\core\controller->executeActions()
#16 E:\leantime\app\core\class.frontcontroller.php(173): leantime\core\controller->__construct()
#17 E:\leantime\app\core\class.frontcontroller.php(233): leantime\core\frontcontroller::executeAction()
#18 E:\leantime\public\theme\default\layout\app.php(35): leantime\core\frontcontroller::includeAction()
#19 E:\leantime\app\core\class.template.php(234): require('...')
#20 E:\leantime\app\domain\dashboard\controllers\class.show.php(94): leantime\core\template->display()
#21 E:\leantime\app\core\class.controller.php(61): leantime\domain\controllers\show->get()
#22 E:\leantime\app\core\class.controller.php(33): leantime\core\controller->executeActions()
#23 E:\leantime\app\core\class.frontcontroller.php(173): leantime\core\controller->__construct()
#24 E:\leantime\app\core\class.frontcontroller.php(95): leantime\core\frontcontroller::executeAction()
#25 E:\leantime\app\core\class.application.php(116): leantime\core\frontcontroller::dispatch()
#26 E:\leantime\public\index.php(26): leantime\core\application->start()
#27 {main}
[26-Jan-2023 08:01:28 America/Los_Angeles] PHP Fatal error:  Cannot declare class leantime\domain\controllers\header, because the name is already in use in E:\leantime\app\domain\pageparts\controllers\class.header.php on line 11
[26-Jan-2023 08:01:29 America/Los_Angeles] PHP Parse error:  syntax error, unexpected 'appSettings' (T_STRING), expecting function (T_FUNCTION) or const (T_CONST) in E:\leantime\app\core\class.application.php on line 12

**Comment by marcelfolaron** [comment]
Thanks. Seems that the php file info extension is not installed. It might
just be a matter of uncommenting it in your php.ini

extension=php_fileinfo.dll

**Reply by Eaclean** [reply]
I enabled that then enabled the GD library and we have a winner!

It looks great, looking forward to exploring the application.

Thank you very much for your support, first class!

**Comment by marcelfolaron** [comment]
Awesome, glad it works! Let me know if you have any feedback.
